### PR TITLE
Typed signing on Ledger error handling + serialization

### PR DESCRIPTION
**Pull request by fernya**
The happy path was already merged but users tend to alter from the intended use cases :)

- Serialize typed signing with other Ledger operations
- Bubble up errors for further processing when we use typed data signing
- Match Redux and LedgerService state before attempting typed signing

### CODE REVIEW

File: background/services/signing/index.ts [hunk 1]
```diff
@@ -178,27 +178,54 @@ export default class SigningService extends BaseService<Events> {
...
-      type: "success-data",
-      signedData,
-    })
+      this.emitter.emit("signingDataResponse", {
+        type: "error",
+        reason: "genericError",
+      })
 
-    return signedData
+      throw err
```
Comment (cairomassimo) [inline background/services/signing/index.ts:227]: Also somewhat duplicated IIRC. This is harder to deduplicate though.

Here is my suggestion:

```typescript
catch (err) {
  this.emitter.emit("signingDataResponse", {
    type: "error",
    reason: getSigningErrorReason(err),
  })
  throw err
}
```

and

```typescript
function getSigningErrorReason(err: unknown): SigningErrorReason {
  if (err instanceof TransportStatusError) {
    const transportError = err as Error & { statusCode: number }
    switch (transportError.statusCode) {
      case StatusCodes.CONDITIONS_OF_USE_NOT_SATISFIED:
        return "userRejected"
      default:
    }
  }

  return "genericError"
}
```

**Review by cairomassimo — CHANGES_REQUESTED** [review]
Approach is good, but we should deduplicate some code before merging (see comments).

File: background/services/ledger/index.ts [hunk 2]
```diff
@@ -384,56 +384,67 @@ export default class LedgerService extends BaseService<Events> {
...
+        !accountData ||
+        path !== accountData.path ||
+        deviceID !== accountData.ledgerId
+      ) {
+        throw new Error("Signing method mismatch!")
+      }
+
+      if (deviceID !== this.#currentLedgerId) {
+        throw new Error("Cannot sign on wrong device attached!")
+      }
```
[the remark this answers is not in the document]
Comment (fernya) [reply background/services/ledger/index.ts:426]: Deduplication in progress!

[the remark this answers is not in the document]
Comment (fernya) [reply background/services/ledger/index.ts]: Done.

File: background/services/signing/index.ts (hunk 1 above)
Comment (fernya) [reply background/services/signing/index.ts]: I followed your suggestion with the necessary bits added by my own.

**Comment by cairomassimo** [comment]
@fernya Great! I'm testing this now

**Review by cairomassimo — APPROVED** [review]
Tested!

**Pull request merged**